ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/blockheaders/NewBlockHeadersSubscriptionServiceTest.java

```diff
@@ -63,76 +73,115 @@
+      new KeyValueStoragePrefixedKeyBlockchainStorage(
+          new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions());
+  private final Block genesisBlock = gen.genesisBlock();
+  private final MutableBlockchain blockchain =
+      DefaultBlockchain.createMutable(genesisBlock, blockchainStorage, new NoOpMetricsSystem());
+
+  @Spy
+  private BlockchainQueries blockchainQueriesSpy =
+      Mockito.spy(new BlockchainQueries(blockchain, createInMemoryWorldStateArchive()));
 
   @Before
   public void before() {
     newBlockHeadersSubscriptionService =
-        new NewBlockHeadersSubscriptionService(subscriptionManager, blockchainQueries);
+        new NewBlockHeadersSubscriptionService(subscriptionManager, blockchainQueriesSpy);
   }
 
   @Test
   public void shouldSendMessageWhenBlockAddedOnCanonicalChain() {
     final NewBlockHeadersSubscription subscription = createSubscription(false);
     mockSubscriptionManagerNotifyMethod(subscription);
-    final BlockResult expectedNewBlock = expectedBlockWithTransactions(Collections.emptyList());
 
-    simulateAddingBlockOnCanonicalChain();
+    final Block testBlock = appendBlockWithParent(blockchain, blockchain.getChainHeadBlock());
+    simulateAddingBlockOnCanonicalChain(testBlock);
+    final BlockResult expectedNewBlock =
+        blockResultFactory.transactionHash(
+            blockchainQueriesSpy.blockByHashWithTxHashes(testBlock.getHash()).orElse(null));
 
     verify(subscriptionManager)
         .sendMessage(subscriptionIdCaptor.capture(), responseCaptor.capture());
     assertThat(subscriptionIdCaptor.getValue()).isEqualTo(subscription.getSubscriptionId());
-    assertThat(responseCaptor.getValue())
-        .isEqualToComparingFieldByFieldRecursively(expectedNewBlock);
+    assertThat(responseCaptor.getValue()).usingRecursiveComparison().isEqualTo(expectedNewBlock);
   }
 
   @Test
   public void shouldNotSendMessageWhenBlockAddedIsNotOnCanonicalChain() {
     simulateAddingBlockOnNonCanonicalChain();
 
-    verifyZeroInteractions(subscriptionManager);
+    verifyNoInteractions(subscriptionManager);
+  }
+
+  @Test
+  public void shouldSendMessagesWhenReorgBlockAdded() {
+    final NewBlockHeadersSubscription subscription = createSubscription(false);
+    mockSubscriptionManagerNotifyMethod(subscription);
+
+    appendBlockWithParent(blockchain, genesisBlock);
+    final Block forkBlock = appendBlockWithParent(blockchain, genesisBlock);
+    final Block forkBlock1 = appendBlockWithParent(blockchain, forkBlock);
+    final Block forkBlock2 = appendBlockWithParent(blockchain, forkBlock1);
+    simulateAddingReorgBlock(forkBlock2, genesisBlock);
+    final BlockResult expectedNewBlock =
+        blockResultFactory.transactionHash(
+            blockchainQueriesSpy.blockByHashWithTxHashes(forkBlock2.getHash()).orElse(null));
+
+    verify(subscriptionManager, times(3)).notifySubscribersOnWorkerThread(any(), any(), any());
+    verify(subscriptionManager, times(3))
```

Comment:
Does this make sure the reorg blocks are sent in the right order?

Reply:
Updated to ensure proper ordering.